Add sphinx documentation

msimberg opened the pull request:
This is now in a state where one can build it via CMake (wanted to avoid this but it's necessary to extract e.g. CMake options into the documentation). I'd be happy if someone wants to try it out. You need `doxygen`, `python`, `python-sphinx`, `python-sphinx-rtd-theme` and `python-breathe` (on Ubuntu). Then set `-DHPX_WITH_DOCUMENTATION=ON` and build with `make docs`. The output currently goes to `<build_dir>/share/hpx/docs/html` where you'll find an `index.html`. The source files live in `docs/sphinx`.

API output looks decent now, and links from text to API reference mostly work. Doxygen still spits out a ton of warnings though as it does with our current documentation.

Still to do:

- [x] merge changes made in master/boostbook
- [x] Add `*.rst` files to inspect checking (copyright/license)
- [x] add links to `share/hpx`
- [x] add git commit
- [x] update documentation using CircleCI
- [x] remove current documentation...
- [x] build documentation only on master and relase branches

## Any background context you want to provide?

#3264 #3263

hkaiser commented on docs/contributing/documentation.rst:
```diff
@@ -0,0 +1,49 @@
+.. _documentation:
```
Add copyright and license, please

msimberg replied:
Thanks for catching this! I suspect there will be more of these, I still need to go through it all at least once.

hkaiser commented:
Can we create a live version of this (similar to the old docs), please?

msimberg commented:
I absolutely want to have a live version. I will try to get it running via CircleCI (first alongside the old docs, then instead of it).

msimberg commented:
I've added a manually built version to the `gh-pages` branch: [https://stellar-group.github.io/hpx/docs/sphinx/html/](https://stellar-group.github.io/hpx/docs/sphinx/html/). CircleCI has changed its policies on so I can't get this to build currently on CircleCI (see #3234). I've had to disable the jekyll option, which means that the README is not displayed on GitHub pages. I'd like to remove that one in any case as it's not updated automatically, and add any relevant content to the documentation.

msimberg commented:
I've just removed the boostbook documentation from this branch. I've added building and pushing docs to the CircleCI config but haven't tested it yet (still needs some configuration on the CircleCI side). I may have to reopen this with a branch from the main repo (not my fork).

If someone feels bored and would like to help with proofreading (mostly formatting, typos, other weirdness as most of the content is the same as before), or final suggestions/complaints on the structure that would be greatly appreciated (@aserio, @NK-Nikunj?).

msimberg commented:
@NK-Nikunj thanks!

@biddisco, you might want to have a look at [this](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/hpx_runtime_and_resources.html#the-hpx-resource-partitioner). It's a section on the resource partitioner. I'm sure it could be expanded significantly. If you have ideas or comments let me know.

@heller, you might want to look at [this](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/debugging_hpx_applications.html). It's a short section on debugging and sanitizers. Would you like to add something about your `hpx/tools/gdb/hpx.py` script (if it's still up to date)?

heller commented:
@sithhell I think they are talking to you again :)

msimberg commented:
Gah. Sorry for the noise, I did indeed mean @sithhell.

sithhell commented on .circleci/config.yml:
```diff
@@ -134,6 +135,30 @@ jobs:
+      - run:
+          name: Pushing Sphinx Documentation
+          command: |
+              ninja git_docs
```
I think this should only happen on the master branch, otherwise, unstable will be overriden by PRs

msimberg replied:
Yeah, that still needs to be added before merging. Added a todo item in the PR description.

msimberg commented:
@NK-Nikunj, the way I imagine new users reading the documentation is by going to the quick start guide, then getting to [this](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/quickstart.html#hpx-dependencies) section and in most cases they would do `apt install libboost-dev google-perftools libhwloc5`. If that doesn't apply to them I would expect expect them to click [prerequisites](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/building_hpx.html#prerequisites) and then [installing boost/hwloc](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/building_hpx.html#installing-boost) for instructions on how to build them. Then they would continue with [this](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/quickstart.html#building-hpx).

But... this is all in my head in an ideal world. So I'd be interested in knowing what makes it difficult to find those instructions and if there's something we could do to highlight it better. And writing this I realize that for example that `apt install ...` line is nowhere in the documentation. Also, [this](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/building_hpx.html#platform-specific-build-recipes) section is probably quite close to what you'd want but it could do with some structure (and love).

NK-Nikunj commented:
@msimberg What I would have preferred when I first looked at the source code was a set of bash commands that I could simply copy-paste to get a build of HPX up and running. Instead, I had to search for the installation guide on boost, hwloc and tcmalloc (I don't know what I did wrong while installing it, but all my apps gave a segmentation fault once I was done installing tcmalloc). Although, it was not a pain installing them, a step easier would have directly been copy-pasting `apt install` commands.

What I meant when I asked to add installation steps was these `apt install` commands. At first, it might not look like an incremental development, but it helps in the long run for people who're beginning with building HPX. [This link](https://stellar-group.github.io/hpx/docs/sphinx/unstable/html/manual/building_hpx.html#platform-specific-build-recipes) would be an apt place for `apt install` commands, no puns intended ;)

msimberg commented:
@NK-Nikunj thanks, that helps a lot. And I agree that it's not just a small thing. Building HPX is not terribly hard in the end, but can become really frustrating if you don't know where to start.

Would you be interested in updating that section (can be in a separate PR once this goes in)? The "How to install HPX on Unix variants" section could contain subsections for what packages to install on Debian/Ubuntu/Fedora/Arch etc. The steps after that should be the same no matter what Linux you're on. Then the existing Fedora section at the end wouldn't be needed anymore, at least in its current form.

NK-Nikunj commented:
@msimberg yes, I can update the section. I think I know exactly what to add as well. However, I'm currently a bit busy, I can start working on it next week if that's not a problem.

msimberg commented:
@NK-Nikunj yeah, that's perfectly fine. We're just happy that you want to help!

msimberg commented:
This would be ready to go in now, modulo any mistakes I might've made. It would be nice to hear if someone other than me has managed to build the documentation successfully, even though CircleCI also managed to do it.

StellarBot on Rostam would have to be disabled once this is merged.

hkaiser commented:
> StellarBot on Rostam would have to be disabled once this is merged.

@hapoo including Alireza...

NK-Nikunj commented:
@msimberg I builds successfully on my PC. It does get a lot of warnings though.

msimberg commented:
@hkaiser @hapoo Merging so that I can check that all is fine on CircleCI. The builder on rostam *should* do nothing now (i.e. `make git_docs` does nothing unless a branch or tag is set through CMake), so there's no rush to disable that. I guess I can disable it as well?

hkaiser commented:
@msimberg thanks! What should we do to make rostam build the new docs instead?

msimberg commented:
@hkaiser simply remove it from rostam :) CircleCI should take care of it from now on.

hkaiser commented:
@msimberg is the release_procedure document gone now?

msimberg commented:
It's still there, it's just been moved to be part of the documentation itself: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/sphinx/contributing/release_procedure.rst.
